### PR TITLE
Injects a client is ready to proceed pre-check

### PR DESCRIPTION
**Pull request by zegnus**
### Problem description

The current DownloadManager has a set of internal rules that specifies if a download can proceed or not, but there is no any mechanism in place to enhance the check following a set of rules defined by a client.

As for example, don't download if the client is not logged in.

### Solution implemented

A client can now specify if the downloads are ready to proceed or not by implementing the interface `DownloadClientReadyChecker` in the Application Activity.

As for example:

```java
public class DemoParallelApplication extends Application implements DownloadClientReadyChecker {
    @Override
    public boolean isReadyToDownload() {
        // rule 1
        // rule 2
        // return rule1IsTrue AND rule2IsTrue
    }
}
```

For now, `isReadyToDownload` don't take any parameters. If that is needed another Pull Request will be issued. Meaning that this method is a global status query, not for a particular download.

### CODE REVIEW

File: library/build.gradle [hunk 1]
```diff
@@ -37,6 +37,10 @@ android {
     lintOptions {
         abortOnError false
     }
+
+    testOptions {
+        unitTests.returnDefaultValues = true
```
Comment (zegnus) [inline library/build.gradle]: Otherwise the test that executes

```
ContentValues values = new ContentValues();
values.put(Downloads.Impl.COLUMN_STATUS, mStatus);
```

will fail

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadClientReadyChecker.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+public interface DownloadClientReadyChecker {
+    Ready READY = new Ready();
+
+    boolean isReadyToDownload();
+
+    /**
+     * Ready specifies that it is always ready to download
+     */
+    class Ready implements DownloadClientReadyChecker {
```
Comment (amlcurran) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadClientReadyChecker.java]: nice

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java [hunk 3]
```diff
@@ -40,10 +40,27 @@ public Reader(ContentResolver resolver, Cursor cursor) {
...
-        public DownloadInfo newDownloadInfo(Context context, SystemFacade systemFacade, StorageManager storageManager, DownloadNotifier notifier) {
-            final DownloadInfo info = new DownloadInfo(context, systemFacade, storageManager, notifier);
+        public DownloadInfo newDownloadInfo(
+                Context context,
+                SystemFacade systemFacade,
+                StorageManager storageManager,
+                DownloadNotifier notifier,
+                DownloadClientReadyChecker downloadClientReadyChecker) {
+
+            Fuzz fuzz = new Fuzz();
```
Comment (amlcurran) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java]: ??

Comment (amlcurran) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java]: as in, what is this ha

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadService.java [hunk 4]
```diff
@@ -154,6 +157,13 @@ public void onCreate() {
         mExecutor = factory.createExecutor();
     }
 
+    private DownloadClientReadyChecker getDownloadClientReadyChecker() {
+        if (!(getApplication() instanceof DownloadClientReadyChecker)) {
```
Comment (amlcurran) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadService.java]: Can't you pass this into the constructor of the DownloadManager? Having a hard dependency on the Application class isn't very extensible

File: library/build.gradle (hunk 1 above)
Comment (blundell) [inline library/build.gradle]: :x: this option should never be turned on.  It means you're testing things not under our control.

Can discuss later :x:

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java [hunk 5]
```diff
@@ -217,7 +239,10 @@ private Long getLong(String column) {
     public int mBypassRecommendedSizeLimit;
     public String bigPictureResourceUrl;
 
-    public int mFuzz;
+    public Uri mAllDownloadsUri;
```
Comment (blundell) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java]: if you're creating new fields, use Novoda coding standards - we are moving away from the Google code and this helps highlight what we have added or changed.

**Comment by xrigau** [comment]
The PR description doesn't explain what happens in case that method returns `false` - does it retry later? does this happen for every download? maybe this is like a `GlobalDownloadClientReadyChecker` and then if you wanted to do it for each download then it could be a `DownloadClientReadyChecker` ?

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java [hunk 6]
```diff
@@ -439,14 +473,14 @@ private NetworkState checkSizeAllowedForNetwork(int networkType) {
...
-            final boolean isReady = isReadyToDownload();
+            final boolean isReady = isDownloadManagerReadyToDownload() && isClientIsReadyToDownload();
             final boolean isActive = mSubmittedTask != null && !mSubmittedTask.isDone();
             if (isReady && !isActive) {
                 if (mStatus != Downloads.Impl.STATUS_RUNNING) {
                     mStatus = Downloads.Impl.STATUS_RUNNING;
                     ContentValues values = new ContentValues();
                     values.put(Downloads.Impl.COLUMN_STATUS, mStatus);
-                    mContext.getContentResolver().update(getAllDownloadsUri(), values, null, null);
+                    mContext.getContentResolver().update(mAllDownloadsUri, values, null, null);
```
Comment (blundell) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java]: why this change?

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java (hunk 3 above)
Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java]: Yeah I had the same question.

If you look down the code at https://github.com/novoda/download-manager/pull/29/files#diff-68db38a7048316539cb0931793be1808L242 you will see that is a random value used in order to avoid clashing for a restart time queries.

It is needed to be injected because otherwise when the test calls https://github.com/novoda/download-manager/blob/master/library/src/main/java/com/novoda/downloadmanager/lib/Helpers.java#L39, the code fails

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java [hunk 7]
```diff
@@ -480,11 +518,12 @@ public boolean isOnCache() {
     }
 
     public Uri getMyDownloadsUri() {
-        return ContentUris.withAppendedId(Downloads.Impl.CONTENT_URI, mId);
+        return mAllDownloadsUri;
```
Comment (blundell) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java]: why this change?

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java [hunk 8]
```diff
@@ -115,6 +132,11 @@ private void addHeader(DownloadInfo info, String header, String value) {
             info.mRequestHeaders.add(Pair.create(header, value));
         }
 
+        private void setDownloadUris(DownloadInfo info) {
+            info.mAllDownloadsUri = ContentUris.withAppendedId(Downloads.Impl.ALL_DOWNLOADS_CONTENT_URI, info.mId);
```
Comment (amlcurran) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java]: why do we need to do this out of curiosity/what's changed?

File: library/src/test/java/com/novoda/downloadmanager/lib/DownloadInfoTest.java [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    public void setUp() {
+        initMocks(this);
+
+        when(mockFuzz.getFuzz()).thenReturn(1);
+        when(mockContext.getContentResolver()).thenReturn(mockContentResolver);
+        when(mockContentResolver.update(any(Uri.class), any(ContentValues.class), any(String.class), any(String[].class))).thenReturn(0);
+    }
+
+    @Test
+    public void givenClientIsReadyToDownloadThenStartDownload() {
```
Comment (amlcurran) [inline library/src/test/java/com/novoda/downloadmanager/lib/DownloadInfoTest.java]: :+1: nice tests

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java (hunk 3 above)
Comment (amlcurran) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java]: OK, perhaps rename or javadoc to make it more obvious?

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadService.java (hunk 4 above)
Comment (ronocod) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadService.java]: :+1: for not forcing it on the Application

File: library/build.gradle (hunk 1 above)
Comment (zegnus) [inline library/build.gradle]: Yup I know, but I didn't want to change too much the code.

The only way is to inject that ContentValues then... or a better idea?

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java (hunk 3 above)
Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java]: :+1:

Comment (ronocod) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java]: Now might be a good time to rename it to something like `RandomNumberGenerator`?

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadService.java (hunk 4 above)
Comment (ronocod) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadService.java]: But it is very difficult to pass dependencies into the `DownloadService` :disappointed:

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java (hunk 7 above)
Comment (ronocod) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java]: Should this be `mMyDownloadsUri`?

**Comment by zegnus** [comment]
@xrigau What happens is that the download don't proceed. As no any logic has been added or modified regarding what to do when a download cannot proceed, I haven't documented it.

There is already an internal *can the download proceed* mechanism, the logic of this new check is added on top of the previous check, and the result is the same.

As a reference, the `activeDownload` will be false here https://github.com/novoda/download-manager/blob/master/library/src/main/java/com/novoda/downloadmanager/lib/DownloadService.java#L344 from https://github.com/novoda/download-manager/blob/master/library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java#L442 where `isReadyToDownload` becoming false

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java (hunk 6 above)
Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java]: Can use the method, reverting

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java (hunk 7 above)
Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java]: Access to `ContentUris` will fail at test. 

As well there is no need to calculate the value every time as it can be precomputed

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java (hunk 8 above)
Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java]: https://github.com/novoda/download-manager/pull/29#discussion_r32744701

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java (hunk 3 above)
Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java]: Yes I thought of that but because that comes from Google I wanted to maintain the consistency of the name just in case we have to merge back again from them, so we know what is what.

Can rename if we don't really care about merging again in the future?

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadService.java (hunk 4 above)
Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadService.java]: Yes exactly, ideas welcome :)

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java (hunk 3 above)
Comment (amlcurran) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java]: Think we said we weren't going to take any more of their code ( #22 )

Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java]: :+1:

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadClientReadyChecker.java (hunk 2 above)
Comment (Electryc) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadClientReadyChecker.java]: :+1:

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java [hunk 10]
```diff
@@ -456,6 +495,10 @@ public boolean startDownloadIfReady(ExecutorService executor) {
         }
     }
 
+    private boolean isClientIsReadyToDownload() {
```
Comment (Electryc) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java]: This reads a little bit odd. Shouldn't it be `isClientReadyToDownload`?

File: library/src/main/java/com/novoda/downloadmanager/lib/RandomNumberGenerator.java [hunk 11]
```diff
@@ -0,0 +1,11 @@
+package com.novoda.downloadmanager.lib;
+
+/**
+ * This class generates a random value to be used as a randomized restart time
+ * for the server queries
+ */
+class RandomNumberGenerator {
+    int getFuzz() {
```
Comment (Electryc) [inline library/src/main/java/com/novoda/downloadmanager/lib/RandomNumberGenerator.java]: `generate()` maybe?

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java (hunk 10 above)
Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java]: yes!

File: demo_parallel/src/main/java/com/novoda/downloadmanager/demo/parallel/DemoParallelApplication.java [hunk 12]
```diff
@@ -0,0 +1,12 @@
+package com.novoda.downloadmanager.demo.parallel;
+
+import android.app.Application;
+
+import com.novoda.downloadmanager.lib.DownloadClientReadyChecker;
+
+public class DemoParallelApplication extends Application implements DownloadClientReadyChecker {
```
Comment (blundell) [inline demo_parallel/src/main/java/com/novoda/downloadmanager/demo/parallel/DemoParallelApplication.java]: In the readme we explicitly wrote that the example for custom rules was in the serial demo not the parralel. No need to add this

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java [hunk 13]
```diff
@@ -40,10 +40,29 @@ public Reader(ContentResolver resolver, Cursor cursor) {
...
+                Context context,
+                SystemFacade systemFacade,
+                StorageManager storageManager,
+                DownloadNotifier notifier,
+                DownloadClientReadyChecker downloadClientReadyChecker) {
+
+            RandomNumberGenerator randomNumberGenerator = new RandomNumberGenerator();
+            ContentValues contentValues = new ContentValues();
+
+            final DownloadInfo info = new DownloadInfo(
```
Comment (blundell) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java]: no need for final

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadService.java (hunk 4 above)
Comment (blundell) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadService.java]: yeah @amlcurran we did look. We don't want to do it reflectively and this service can be started from a broadcast receiver when the only thing that is available is the Application.

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java [hunk 14]
```diff
@@ -115,6 +134,11 @@ private void addHeader(DownloadInfo info, String header, String value) {
             info.mRequestHeaders.add(Pair.create(header, value));
         }
 
+        private void setDownloadUris(DownloadInfo info) {
+            info.allDownloadsUri = ContentUris.withAppendedId(Downloads.Impl.ALL_DOWNLOADS_CONTENT_URI, info.mId);
```
Comment (blundell) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java]: :x: this change breaks the current behaviour.

Let's try and do minimal changes to the DownloadManager or be extra careful when doing them.

This breaks the behaviour as now you "pre-compute the allDownloadsUri" however if the `info.mId` changes which it can! see method `updateFromDatabase()`

**Comment by blundell** [comment]
LGTM :+1: :shipit:

**Comment by blundell** [comment]
ah yes you're right, my local build broke but I ran out of laptop battery, I'm on it :+1:

**Comment by blundell** [comment]
@zegnus the problem was you where testing something that wasn't isolated, the downloadmanager code is not architectured with tests in mind and sometimes to change this is close to breaking behaviour or introducing bugs that we should not be doing, it's a tight rope to walk and we have to be careful. Minimal changes recommended at all times.

**Comment by stefanhoth** [comment]
`DownloadClientReadyChecker.isReadyToDownload` dat naming :eyes:

**Comment by zegnus** [comment]
@blundell Yup, I had the same feeling. I wanted to start to write tests on this module because there are almost none but as you said this is not encoded in a way that helps testing, at all and I wanted to the the minimal changes for the tests.

I thought about the same as you did here https://github.com/novoda/download-manager/commit/f3b4b6aad0ee816a83c0fff067c33fb05c9a17c7 didn't want to change much. But considering how the dependencies of the download method mixes up different concepts and makes the tests a bit of a hell, this is much much better :+1:

**Comment by xrigau** [comment]
@zegnus :+1:

**Comment by xrigau** [comment]
no outstanding issues so I say :shipit: and iterate :+1:
